File: server/java/src/main/java/com/server/reveal/UserContextProvider.java
package com.server.reveal;

import io.revealbi.core.IRVUserContext;
import io.revealbi.core.RVUserContext;
import io.revealbi.servlet.IRVServletUserContextProvider;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.HashMap;

@Component
public class UserContextProvider implements IRVServletUserContextProvider {

    @Value("${POSTGRES_HOST:localhost}")
    private String postgresHost;

    @Value("${POSTGRES_DATABASE:Northwind}")
    private String postgresDatabase;

    @Value("${POSTGRES_USERNAME:postgres}")
    private String postgresUsername;

    @Value("${POSTGRES_PASSWORD:}")
    private String postgresPassword;

    @Value("${POSTGRES_SCHEMA:public}")
    private String postgresSchema;

    @Override
    public IRVUserContext getUserContext(HttpServletRequest request) {
        String headerValue = request.getHeader("x-header-one");
        String userId = null;
        String orderId = null;

        if (headerValue != null && !headerValue.isEmpty()) {
            String[] pairs = headerValue.split(",");
            for (String pair : pairs) {
                String[] kv = pair.split(":", 2);
                if (kv.length == 2) {
                    String key = kv[0].trim();
                    String value = kv[1].trim();
                    if (key.equalsIgnoreCase("userId")) {
                        userId = value;
                    } else if (key.equalsIgnoreCase("orderId")) {
                        orderId = value;
                    }
                }
            }
        }

        // default to User role
        String role = "User";

        // null is used here just for demo
        if ("BLONP".equals(userId) || userId == null) {
            role = "Admin";
        }

        String[] filterTables = role.equals("Admin")
            ? new String[0]
            : new String[]{"customers", "orders"};

        var props = new HashMap<String, Object>();
        props.put("OrderId", orderId != null ? orderId : "");
        props.put("Role", role);
        props.put("Host", postgresHost);
        props.put("Database", postgresDatabase);
        props.put("Username", postgresUsername);
        props.put("Password", postgresPassword);
        props.put("Schema", postgresSchema);
        props.put("FilterTables", filterTables);

        return new RVUserContext(userId, props);
    }
}

File: server/aspnet/RevealSdk.Server/Reveal/UserContextProvider.cs
using Microsoft.Extensions.Options;
using Reveal.Sdk;
using RevealSdk.Server.Configuration;

namespace RevealSdk.Server.Reveal
{

    public class UserContextProvider : IRVUserContextProvider
    {
        private readonly ServerOptions _options;

        public UserContextProvider(IOptions<ServerOptions> sqlOptions)
        {
            _options = sqlOptions.Value;
        }

        IRVUserContext IRVUserContextProvider.GetUserContext(HttpContext aspnetContext)
        {
            string? headerValue = aspnetContext.Request.Headers["x-header-one"].FirstOrDefault();
            string? userId = null;
            string? orderId = null;

            if (!string.IsNullOrEmpty(headerValue))
            {
                var pairs = headerValue.Split(',');
                foreach (var pair in pairs)
                {
                    var kv = pair.Split(':', 2);
                    if (kv.Length == 2)
                    {
                        var key = kv[0].Trim();
                        var value = kv[1].Trim();
                        if (key.Equals("userId", StringComparison.OrdinalIgnoreCase))
                            userId = value;
                        else if (key.Equals("orderId", StringComparison.OrdinalIgnoreCase))
                            orderId = value;
                    }
                }
            }

            // default to User role
            string role = "User";

            // null is used here just for demo 
            if (userId == "BLONP" || userId == null)
            {
                role = "Admin";
            }

            var filterTables = role == "Admin"
                ? Array.Empty<string>()
                : ["customers", "orders"];

            // Use hardcoded host if _options.Host is empty
            string host = string.IsNullOrEmpty(_options.Host)
                ? ""
                : _options.Host;

            var props = new Dictionary<string, object>() {
                { "OrderId", orderId ?? string.Empty },
                { "Role", role },
                { "Host", host },
                { "Database", _options.Database },
                { "Username", _options.Username },
                { "Password", _options.Password },
                { "Schema", _options.Schema },
                { "Port", _options.Port },
                { "FilterTables", filterTables }
            };

            return new RVUserContext(userId, props);
        }
    }
}